csharp/src/Ice/Proxy.cs

```diff
@@ -252,6 +269,80 @@ public static class Proxy
             CancellationToken cancel = default) =>
             InvokeWithInterceptorsAsync(proxy, request, oneway, synchronous: false, progress, cancel);
 
+        /// <summary>Sends a request that returns a value and returns the result asynchronously.</summary>
+        /// <typeparam name="T">The operation's return type.</typeparam>
+        /// <param name="proxy">The proxy for the target Ice object.</param>
+        /// <param name="request">The <see cref="OutgoingRequestFrame"/> for this invocation. Usually this request
+        /// frame should have been created using the same proxy, however some differences are acceptable, for example
+        /// proxy can have different endpoints.</param>
+        /// <param name="reader">An <see cref="InputStreamReader{T}"/> for the operation's return value. Typically
+        /// {IInterfaceNamePrx}.Response.{OperationName}.</param>
+        /// <param name="progress">Sent progress provider.</param>
+        /// <param name="cancel">A cancellation token that receives the cancellation requests.</param>
+        /// <returns>The return value.</returns>
+        public static Task<T> InvokeAsync<T>(
+            this IObjectPrx proxy,
+            OutgoingRequestFrame request,
+            InputStreamReader<T> reader,
+            IProgress<bool>? progress = null,
+            CancellationToken cancel = default)
+        {
+            return ReadResponseAsync(proxy.InvokeAsync(request, oneway: false, progress, cancel),
+                                     reader,
+                                     proxy.Communicator);
+
+            static async Task<T> ReadResponseAsync(
+                ValueTask<IncomingResponseFrame> task,
+                InputStreamReader<T> reader,
+                Communicator communicator) =>
+                (await task.ConfigureAwait(false)).ReadReturnValue(communicator, reader);
```

Comment:
Why a separate method to read the response? This could just be:
```
        public static async Task<T> InvokeAsync<T>(
            this IObjectPrx proxy,
            OutgoingRequestFrame request,
            InputStreamReader<T> reader,
            IProgress<bool>? progress = null,
            CancellationToken cancel = default) =>
            (await proxy.InvokeAsync(request, 
                                     oneway: false, 
                                     progress, 
                                     cancel).ConfigureAwait(false)).ReadReturnValue(proxy.Communicator, reader);
```

Reply:
It's because we don't want to mark the main method (InvokeAsync) "async", so that exceptions thrown during the synchronous preamble of this method are thrown synchronously to the caller (and not held in the returned task).

This synchronous preamble includes all the invocation interceptors that can execute synchronously. See the existing public InvokeAsync and its implementation.

Reply:
Ok. Could we instead use non-static embedded methods to simplify a bit the code?